proxy/proxy/client_test.go

```diff
@@ -183,3 +192,43 @@ func TestMaximumConnectionsCount(t *testing.T) {
 		t.Errorf("client should have dialed exactly the maximum of %d connections (%d connections, %d dials)", maxConnections, numConnections, dials)
 	}
 }
+
+func TestRefreshTimer(t *testing.T) {
+	refreshCertBuffer = time.Second
+	timeToExpire := 2 * time.Second
+	b := &fakeCerts{}
+	c := &Client{
+		Certs: &blockingCertSource{
+			map[string]*fakeCerts{
+				instance: b,
+			},
+			time.Now().Add(timeToExpire),
+		},
+		Dialer: func(string, string) (net.Conn, error) {
+			return nil, errFakeDial
+		},
+		RefreshCfgThrottle: 500 * time.Millisecond,
+	}
+
+	// Call Dial to cache the cert.
+	if _, err := c.Dial(instance); err != errFakeDial {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	cached, ok := c.cfgCache[instance]
+	if !ok {
+		t.Error("expected instance to be cached")
+	}
+
+	// Wait for cert to expire.
+	time.Sleep(timeToExpire + time.Second)
```

Comment:
In general we should avoid sleeping in tests, it tends to add flakiness to otherwise deterministic tests. But in this case we're waiting for something in the background to finish so I think it's OK. But to cut down on flakiness, I suggest looping a few times (with short sleeps in the middle) in case the threads are getting scheduled in an interesting order.